core/src/main/java/org/elasticsearch/common/settings/AbstractScopedSettings.java

```diff
@@ -44,6 +45,9 @@
     private final Map<String, Setting<?>> complexMatchers = new HashMap<>();
     private final Map<String, Setting<?>> keySettings = new HashMap<>();
     private final Setting.Scope scope;
+    private static final Pattern KEY_PATTERN = Pattern.compile("^([\\w\\d_-]+[.])*[\\w\\d_-]+$");
```

Comment:
I think this should be:

    ^(?:[-\\w]+[.])*[-\\w]+$

* non-capturing groups (`(?:..)`)  are more efficient
* `\w` already includes `\d` and `_`
* a `-` inside `[  ]` should appear first, otherwise it indicates a range (at least in pcre)